test/expr/delayed-ident/member_chains.swift

```diff
@@ -0,0 +1,211 @@
+let _: ImplicitMembers = .implicit.another[optional: ()]!
+let _: ImplicitMembers? = .implicit.another[optional: ()]
+let _: ImplicitMembers = .implicit.another[func: ()]()
+let _: ImplicitMembers = .implicit.another[funcOptional: ()]()!
+let _: ImplicitMembers? = .implicit.another[funcOptional: ()]()
+let _: ImplicitMembers = .implicit.another[optionalFunc: ()]!()
+let _: ImplicitMembers? = .implicit.another[optionalFunc: ()]?()
+let _: ImplicitMembers = .implicit[()].another
+let _: ImplicitMembers = .implicit[optional: ()]!.another
+let _: ImplicitMembers? = .implicit[optional: ()]?.another
+let _: ImplicitMembers = .implicit[func: ()]().another
+let _: ImplicitMembers = .implicit[funcOptional: ()]()!.another
+let _: ImplicitMembers? = .implicit[funcOptional: ()]()?.another
+let _: ImplicitMembers = .implicit[optionalFunc: ()]!().another
+let _: ImplicitMembers? = .implicit[optionalFunc: ()]?().another
+
+func implicit(_ i: inout ImplicitMembers) {
+    if i == .implicit {}
+    if i == .implicit.another {}
+    if i == .implicit.getAnother() {}
+    if i == .optional?.another {}
+    if i == .optional!.another {}
+    if i == .createOptional()?.another {}
+}
+
+struct ImplicitGeneric<T> {
+    static var implicit: ImplicitGeneric<T> { ImplicitGeneric<T>() }
+    var another: ImplicitGeneric<T> { ImplicitGeneric<T>() }
+    func getAnother() -> ImplicitGeneric<T> {
+        ImplicitGeneric<T>()
+    }
+}
+
+extension ImplicitGeneric where T == Int {
+    static var implicitInt: ImplicitGeneric<Int> { ImplicitGeneric<Int>() }
+    static var implicitString: ImplicitGeneric<String> { ImplicitGeneric<String>() }
+    var anotherInt: ImplicitGeneric<Int> { ImplicitGeneric<Int>() }
+    var anotherIntString: ImplicitGeneric<String> { ImplicitGeneric<String>() }
+    func getAnotherInt() -> ImplicitGeneric<Int> {
+        ImplicitGeneric<Int>()
+    }
+}
+
+extension ImplicitGeneric where T == String {
+    static var implicitString: ImplicitGeneric<String> { ImplicitGeneric<String>() }
+    var anotherString: ImplicitGeneric<String> { ImplicitGeneric<String>() }
+    var anotherStringInt: ImplicitGeneric<Int> { ImplicitGeneric<Int>() }
+    func getAnotherString() -> ImplicitGeneric<String> {
+        ImplicitGeneric<String>()
+    }
+}
+
+func implicit<T>(_ arg: ImplicitGeneric<T>) {}
+
+implicit(.implicitInt)
+implicit(.implicit.anotherInt)
+implicit(.implicit.anotherInt.another)
+implicit(.implicit.another.anotherInt)
+implicit(.implicit.getAnotherInt())
+implicit(.implicit.another.getAnotherInt())
+implicit(.implicit.getAnother().anotherInt)
+implicit(.implicit.getAnotherInt())
+implicit(.implicit.getAnother().getAnotherInt())
+implicit(.implicitString.anotherStringInt)
+// Member types along the chainc can have different generic arguments
+implicit(.implicit.anotherIntString.anotherStringInt)
+
+implicit(.implicit.anotherString.anotherStringInt) // expected-error {{type of expression is ambiguous without more context}}
```

Comment:
I know this is the initial implementation, but this diagnostic could be improved to something like `cannot infer contextual base in reference to member chain '...'` :)

Reply:
Yeah, improving the diagnostic here is definitely on the TODO list...

Reply:
This now reads `member chain produces result of type 'ImplicitGeneric<Int>' but contextual base was inferred as 'ImplicitGeneric<String>'` with a note at `struct ImplicitGeneric<T> {` of `arguments to generic parameter 'T' ('Int' and 'String') are expected to be equal`.